Modify Time class in Particle class

aysunrhn opened the pull request:
I modified the Time class in the SedTRAILS internal Particle class following our discussions about how to incorporate time internally. Closes #160 

## Important points to note:
- We will stick to the `numpy.datetime64` in seconds to represent time throughout the simulation
- The time difference (or time offset, or simulation run time) is a `numpy.timedelta64` object in seconds
- Also added preliminary tests for the modified Time class in `test/particle_tracer/test_particle.py` as a new `TestTime` class

## Relevant files:
- `src/sedtrails/particle_tracer/particle.py`
- `src/sedtrails/tests/particle_tracer/test_particle.py`

aysunrhn committed:
Update particle.py to modify Time class

aysunrhn committed:
Add tests for the updated Test class in Particle class

manuGil commented on src/sedtrails/particle_tracer/particle.py:
```diff
@@ -16,52 +16,57 @@ class Time:
     ----------
-    time_step : int
-        The current time step in a simulation. A non-negative integer.
+    reference_date : numpy.datetime64
```
Could you modify the datatype of `reference_date` so that we pass a string with the date, instead of the numpy object? For example: `reference_date= "2025-05-14"`

manuGil replied:
Then, the class does the transformation that are needed.

manuGil commented on src/sedtrails/particle_tracer/particle.py:
```diff
@@ -16,52 +16,57 @@ class Time:
-        The current time step in a simulation. A non-negative integer.
+    reference_date : numpy.datetime64
+        The reference date for calculating time.
+    offset_seconds : numpy.timedelta64
```
Would be more intuitive to pass pass only the number of units to `offset_seconds` and hard code the units (seconds). That way it is easier for humans to interpret. I am thinking that in the configuration file, something like the size of the time step will be defined in seconds, for example. We could also accept both data types, might be a case in which the offset can be process faster if in deltatime

aysunrhn replied:
In this case, do you recommend that we make the `offset_seconds` variable either an integer or deltatime object?

manuGil replied:
Let's make it an integer.  The class should  be responsible of converting the type is needed

manuGil commented on src/sedtrails/particle_tracer/particle.py:
```diff
@@ -16,52 +16,57 @@ class Time:
-        if time_step < 0:
-            raise ValueError(f"Expected 'time' to be a non-negative integer, got {time_step}")
-        self.time_step = time_step
+        return self.reference_date + self.offset_seconds
```
Can you double check if this the the correct way to calculate the current simulation time? 
According to my conversion with @bartvanwesten, what they need is a way in which the point from which time is calculated can be changed. For example, having the capability to set at which Gregorian date/time the simulation time starts. That is they need to say simulation time starts at 1 May 2025 midnight, therefore, the current time should be the number of second since that date/time. I have the impression that the current code does not enable that.

manuGil replied:
Can you also add a convenient function to get the time in seconds since the reference date/time

aysunrhn replied:
If I understand correctly, the reference_date is different from the time the simulation starts. I think I confused the two concepts here.

manuGil replied:
That is correct.  The reference-date defines what is time zero on a simulation run. Then a simulation can start at any time after that. 

For example, we want to be able to say,  for simulation S, the reference_date Rd is 16 May, therefore, the simulation time Ts is 0 at RD, and it progresses  from there.  Then, S can have a starting time St that is N number of second  from RD. 

Does this makes things clearer? If not let's have a chat  next week.

aysunrhn replied:
Thank you for looking into the code!

I changed the code to accept `offset_seconds` as int. Now we have a method to convert it to np.deltatime64 (we should check later if this is computationally more efficient way to go).

I also modified the relevant test functions in `tests/particle_tracer/test_particle.py` to be consistent with the new variable convention. 

It will be good to talk tomorrow to clarify some of the details, to make sure I got it correct.

manuGil commented on tests/particle_tracer/test_particle.py:
```diff
@@ -97,3 +98,52 @@ class TestPhysics:
+        assert time_instance.get_current_time() == expected_time, \
+            f"Updated current time should be {expected_time}, but got {time_instance.get_current_time()}."
+
+    def test_update_subtract_seconds(self):
```
simulation time will move very likely only forwards. But it is good check it can be use either way.

manuGil reviewed (changes requested):
@aysunrhn Could you please consider my comments and make changes to the code?

aysunrhn committed:
Accept reference_date as str

Change the reference_date format to str and add a method to convert the reference_date str to a numpy.datetime64 object

aysunrhn committed:
Modify tests to accept reference_date as str

Modified the tests for the Time class to accept reference_date as a str, and added a new test function for the method to retrieve number of seconds since reference_date

aysunrhn committed:
Change offset_seconds to int, add a conversion method and modify the relevant tests

manuGil reviewed (approved):
@aysunrhn  Thank you for adapting the code. I will merge this for now. For further changes we can make a new pull request.
